Fix flakey destroyed_by_association tests

jonathanhefner opened the pull request:
Example failure: https://buildkite.com/rails/rails/builds/68661#84f8790a-fc9e-42ef-a7fb-5bd15a489de8/1002-1012

The failing `destroyed_by_association` tests create an author (a `DestroyByParentAuthor`) and a book (a `DestroyByParentBook`) that belongs to that author.  If the database already contains books that refer to that author's ID from previous tests (i.e. tests that disabled `use_transactional_tests`), then one of those books will be loaded and destroyed instead of the intended `DestroyByParentBook` book.

By loading the `:books` fixtures, we ensure the database does not contain such unexpected books.

kamipo commented:
I can reproduce that failures locally with `bundle exec rake test_xxx`, but `bin/test` is not.

jonathanhefner commented:
@kamipo Were you able to reproduce the failures on both `master` and my branch?

I tried running `bundle exec rake test:sqlite3_mem TEST=test/cases/associations/has_one_associations_test.rb` several times, but I was still not able to reproduce the failures locally.

kamipo commented:
Seem this failure depends on test execution order, not happen with testing `test/cases/associations/has_one_associations_test.rb` only.

kamipo commented:
On master (ea4acf540e07b23cdf51834132f2c60ea5c9a93f):

```
% bundle exec rake test_sqlite3 --verbose TESTOPTS="--seed=29932"
Use TESTOPTS="--verbose" to pass --verbose, etc. to runners.
/Users/kamipo/.rbenv/versions/2.7.0/bin/ruby -w -I"lib:test" -I"/Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.7.0/gems/rake-13.0.1/lib" "/Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.7.0/gems/rake-13.0.1/lib/rake/rake_test_loader.rb" "test/cases/adapter_test.rb" "test/cases/aggregations_test.rb" "test/cases/annotate_test.rb" "test/cases/ar_schema_test.rb" "test/cases/arel/attributes/attribute_test.rb" "test/cases/arel/attributes/math_test.rb" "test/cases/arel/attributes_test.rb" "test/cases/arel/collectors/bind_test.rb" "test/cases/arel/collectors/composite_test.rb" "test/cases/arel/collectors/sql_string_test.rb" "test/cases/arel/collectors/substitute_bind_collector_test.rb" "test/cases/arel/crud_test.rb" "test/cases/arel/delete_manager_test.rb" "test/cases/arel/factory_methods_test.rb" "test/cases/arel/insert_manager_test.rb" "test/cases/arel/nodes/and_test.rb" "test/cases/arel/nodes/as_test.rb" "test/cases/arel/nodes/ascending_test.rb" "test/cases/arel/nodes/bin_test.rb" "test/cases/arel/nodes/binary_test.rb" "test/cases/arel/nodes/bind_param_test.rb" "test/cases/arel/nodes/case_test.rb" "test/cases/arel/nodes/casted_test.rb" "test/cases/arel/nodes/comment_test.rb" "test/cases/arel/nodes/count_test.rb" "test/cases/arel/nodes/delete_statement_test.rb" "test/cases/arel/nodes/descending_test.rb" "test/cases/arel/nodes/distinct_test.rb" "test/cases/arel/nodes/equality_test.rb" "test/cases/arel/nodes/extract_test.rb" "test/cases/arel/nodes/false_test.rb" "test/cases/arel/nodes/grouping_test.rb" "test/cases/arel/nodes/infix_operation_test.rb" "test/cases/arel/nodes/insert_statement_test.rb" "test/cases/arel/nodes/named_function_test.rb" "test/cases/arel/nodes/node_test.rb" "test/cases/arel/nodes/not_test.rb" "test/cases/arel/nodes/or_test.rb" "test/cases/arel/nodes/over_test.rb" "test/cases/arel/nodes/select_core_test.rb" "test/cases/arel/nodes/select_statement_test.rb" "test/cases/arel/nodes/sql_literal_test.rb" "test/cases/arel/nodes/sum_test.rb" "test/cases/arel/nodes/table_alias_test.rb" "test/cases/arel/nodes/true_test.rb" "test/cases/arel/nodes/unary_operation_test.rb" "test/cases/arel/nodes/update_statement_test.rb" "test/cases/arel/nodes/window_test.rb" "test/cases/arel/nodes_test.rb" "test/cases/arel/select_manager_test.rb" "test/cases/arel/table_test.rb" "test/cases/arel/update_manager_test.rb" "test/cases/arel/visitors/dispatch_contamination_test.rb" "test/cases/arel/visitors/dot_test.rb" "test/cases/arel/visitors/mysql_test.rb" "test/cases/arel/visitors/postgres_test.rb" "test/cases/arel/visitors/sqlite_test.rb" "test/cases/arel/visitors/to_sql_test.rb" "test/cases/associations/belongs_to_associations_test.rb" "test/cases/associations/bidirectional_destroy_dependencies_test.rb" "test/cases/associations/callbacks_test.rb" "test/cases/associations/cascaded_eager_loading_test.rb" "test/cases/associations/eager_load_includes_full_sti_class_test.rb" "test/cases/associations/eager_load_nested_include_test.rb" "test/cases/associations/eager_singularization_test.rb" "test/cases/associations/eager_test.rb" "test/cases/associations/extension_test.rb" "test/cases/associations/has_and_belongs_to_many_associations_test.rb" "test/cases/associations/has_many_associations_test.rb" "test/cases/associations/has_many_through_associations_test.rb" "test/cases/associations/has_one_associations_test.rb" "test/cases/associations/has_one_through_associations_test.rb" "test/cases/associations/inner_join_association_test.rb" "test/cases/associations/inverse_associations_test.rb" "test/cases/associations/join_model_test.rb" "test/cases/associations/left_outer_join_association_test.rb" "test/cases/associations/nested_through_associations_test.rb" "test/cases/associations/required_test.rb" "test/cases/associations_test.rb" "test/cases/attribute_decorators_test.rb" "test/cases/attribute_methods/read_test.rb" "test/cases/attribute_methods_test.rb" "test/cases/attributes_test.rb" "test/cases/autosave_association_test.rb" "test/cases/base_test.rb" "test/cases/batches_test.rb" "test/cases/binary_test.rb" "test/cases/bind_parameter_test.rb" "test/cases/boolean_test.rb" "test/cases/cache_key_test.rb" "test/cases/calculations_test.rb" "test/cases/callbacks_test.rb" "test/cases/clone_test.rb" "test/cases/coders/json_test.rb" "test/cases/coders/yaml_column_test.rb" "test/cases/collection_cache_key_test.rb" "test/cases/column_alias_test.rb" "test/cases/column_definition_test.rb" "test/cases/comment_test.rb" "test/cases/connection_adapters/adapter_leasing_test.rb" "test/cases/connection_adapters/connection_handler_test.rb" "test/cases/connection_adapters/connection_handlers_multi_db_test.rb" "test/cases/connection_adapters/connection_handlers_multi_pool_config_test.rb" "test/cases/connection_adapters/connection_handlers_sharding_db_test.rb" "test/cases/connection_adapters/merge_and_resolve_default_url_config_test.rb" "test/cases/connection_adapters/mysql_type_lookup_test.rb" "test/cases/connection_adapters/schema_cache_test.rb" "test/cases/connection_adapters/type_lookup_test.rb" "test/cases/connection_management_test.rb" "test/cases/connection_pool_test.rb" "test/cases/core_test.rb" "test/cases/counter_cache_test.rb" "test/cases/custom_locking_test.rb" "test/cases/database_configurations/hash_config_test.rb" "test/cases/database_configurations/resolver_test.rb" "test/cases/database_configurations_test.rb" "test/cases/database_selector_test.rb" "test/cases/database_statements_test.rb" "test/cases/date_test.rb" "test/cases/date_time_precision_test.rb" "test/cases/date_time_test.rb" "test/cases/defaults_test.rb" "test/cases/dirty_test.rb" "test/cases/disconnected_test.rb" "test/cases/dup_test.rb" "test/cases/enum_test.rb" "test/cases/errors_test.rb" "test/cases/explain_subscriber_test.rb" "test/cases/explain_test.rb" "test/cases/filter_attributes_test.rb" "test/cases/finder_respond_to_test.rb" "test/cases/finder_test.rb" "test/cases/fixture_set/file_test.rb" "test/cases/fixtures_test.rb" "test/cases/forbidden_attributes_protection_test.rb" "test/cases/habtm_destroy_order_test.rb" "test/cases/hot_compatibility_test.rb" "test/cases/i18n_test.rb" "test/cases/inheritance_test.rb" "test/cases/insert_all_test.rb" "test/cases/instrumentation_test.rb" "test/cases/integration_test.rb" "test/cases/invalid_connection_test.rb" "test/cases/invertible_migration_test.rb" "test/cases/json_attribute_test.rb" "test/cases/json_serialization_test.rb" "test/cases/locking_test.rb" "test/cases/log_subscriber_test.rb" "test/cases/migration/change_schema_test.rb" "test/cases/migration/change_table_test.rb" "test/cases/migration/column_attributes_test.rb" "test/cases/migration/column_positioning_test.rb" "test/cases/migration/columns_test.rb" "test/cases/migration/command_recorder_test.rb" "test/cases/migration/compatibility_test.rb" "test/cases/migration/create_join_table_test.rb" "test/cases/migration/foreign_key_test.rb" "test/cases/migration/index_test.rb" "test/cases/migration/logger_test.rb" "test/cases/migration/pending_migrations_test.rb" "test/cases/migration/references_foreign_key_test.rb" "test/cases/migration/references_index_test.rb" "test/cases/migration/references_statements_test.rb" "test/cases/migration/rename_table_test.rb" "test/cases/migration_test.rb" "test/cases/migrator_test.rb" "test/cases/mixin_test.rb" "test/cases/modules_test.rb" "test/cases/multi_db_migrator_test.rb" "test/cases/multiparameter_attributes_test.rb" "test/cases/multiple_db_test.rb" "test/cases/nested_attributes_test.rb" "test/cases/nested_attributes_with_callbacks_test.rb" "test/cases/null_relation_test.rb" "test/cases/numeric_data_test.rb" "test/cases/persistence_test.rb" "test/cases/pooled_connections_test.rb" "test/cases/prepared_statement_status_test.rb" "test/cases/primary_keys_test.rb" "test/cases/query_cache_test.rb" "test/cases/quoting_test.rb" "test/cases/readonly_test.rb" "test/cases/reaper_test.rb" "test/cases/reflection_test.rb" "test/cases/relation/delegation_test.rb" "test/cases/relation/delete_all_test.rb" "test/cases/relation/merging_test.rb" "test/cases/relation/mutation_test.rb" "test/cases/relation/or_test.rb" "test/cases/relation/predicate_builder_test.rb" "test/cases/relation/record_fetch_warning_test.rb" "test/cases/relation/select_test.rb" "test/cases/relation/update_all_test.rb" "test/cases/relation/where_chain_test.rb" "test/cases/relation/where_clause_test.rb" "test/cases/relation/where_test.rb" "test/cases/relation_test.rb" "test/cases/relations_test.rb" "test/cases/reload_models_test.rb" "test/cases/reserved_word_test.rb" "test/cases/result_test.rb" "test/cases/sanitize_test.rb" "test/cases/schema_dumper_test.rb" "test/cases/schema_loading_test.rb" "test/cases/scoping/default_scoping_test.rb" "test/cases/scoping/named_scoping_test.rb" "test/cases/scoping/relation_scoping_test.rb" "test/cases/secure_token_test.rb" "test/cases/serialization_test.rb" "test/cases/serialized_attribute_test.rb" "test/cases/statement_cache_test.rb" "test/cases/statement_invalid_test.rb" "test/cases/store_test.rb" "test/cases/strict_loading_test.rb" "test/cases/suppressor_test.rb" "test/cases/tasks/database_tasks_test.rb" "test/cases/tasks/mysql_rake_test.rb" "test/cases/tasks/postgresql_rake_test.rb" "test/cases/tasks/sqlite_rake_test.rb" "test/cases/test_databases_test.rb" "test/cases/test_fixtures_test.rb" "test/cases/time_precision_test.rb" "test/cases/timestamp_test.rb" "test/cases/touch_later_test.rb" "test/cases/transaction_callbacks_test.rb" "test/cases/transaction_isolation_test.rb" "test/cases/transactions_test.rb" "test/cases/type/adapter_specific_registry_test.rb" "test/cases/type/date_time_test.rb" "test/cases/type/integer_test.rb" "test/cases/type/string_test.rb" "test/cases/type/time_test.rb" "test/cases/type/type_map_test.rb" "test/cases/type/unsigned_integer_test.rb" "test/cases/type_test.rb" "test/cases/types_test.rb" "test/cases/unconnected_test.rb" "test/cases/unsafe_raw_sql_test.rb" "test/cases/validations/absence_validation_test.rb" "test/cases/validations/association_validation_test.rb" "test/cases/validations/i18n_generate_message_validation_test.rb" "test/cases/validations/i18n_validation_test.rb" "test/cases/validations/length_validation_test.rb" "test/cases/validations/numericality_validation_test.rb" "test/cases/validations/presence_validation_test.rb" "test/cases/validations/uniqueness_validation_test.rb" "test/cases/validations_test.rb" "test/cases/view_test.rb" "test/cases/yaml_serialization_test.rb" "test/cases/adapters/sqlite3/collation_test.rb" "test/cases/adapters/sqlite3/copy_table_test.rb" "test/cases/adapters/sqlite3/explain_test.rb" "test/cases/adapters/sqlite3/json_test.rb" "test/cases/adapters/sqlite3/quoting_test.rb" "test/cases/adapters/sqlite3/sqlite3_adapter_test.rb" "test/cases/adapters/sqlite3/sqlite3_create_folder_test.rb" "test/cases/adapters/sqlite3/statement_pool_test.rb" "test/cases/adapters/sqlite3/transaction_test.rb" --seed=29932
Using sqlite3
Run options: --seed=29932

# Running:

.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................S...........................................................................................................................................................................................................................................................................................S..........................................................................................................................................................................................................................................................................................................................................................................................S......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................S................................................................................................................................................................................F

Failure:
HasOneAssociationsTest#test_destroyed_by_association_set_in_child_destroy_callback_on_parent_destroy [/Users/kamipo/src/github.com/rails/rails/activerecord/test/cases/associations/has_one_associations_test.rb:818]:
Expected true to be nil or false


rails test Users/kamipo/src/github.com/rails/rails/activerecord/test/cases/associations/has_one_associations_test.rb:812

..........................F

Failure:
HasOneAssociationsTest#test_destroyed_by_association_set_in_child_destroy_callback_on_replace [/Users/kamipo/src/github.com/rails/rails/activerecord/test/cases/associations/has_one_associations_test.rb:828]:
Expected true to be nil or false


rails test Users/kamipo/src/github.com/rails/rails/activerecord/test/cases/associations/has_one_associations_test.rb:821

.......................................................................................................................................................................................................................................................................S.......................................................................................................................................................................................S.................................................................................................................S...S.......S........S........S...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................S............................................................................................................................................................................S....S....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 131.102285s, 51.2119 runs/s, 137.1143 assertions/s.
6714 runs, 17976 assertions, 2 failures, 0 errors, 14 skips

You have skipped tests. Run with --verbose for details.
rake aborted!
Command failed with status (1): [ruby -w -I"lib:test" -I"/Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.7.0/gems/rake-13.0.1/lib" "/Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.7.0/gems/rake-13.0.1/lib/rake/rake_test_loader.rb" "test/cases/adapter_test.rb" "test/cases/aggregations_test.rb" "test/cases/annotate_test.rb" "test/cases/ar_schema_test.rb" "test/cases/arel/attributes/attribute_test.rb" "test/cases/arel/attributes/math_test.rb" "test/cases/arel/attributes_test.rb" "test/cases/arel/collectors/bind_test.rb" "test/cases/arel/collectors/composite_test.rb" "test/cases/arel/collectors/sql_string_test.rb" "test/cases/arel/collectors/substitute_bind_collector_test.rb" "test/cases/arel/crud_test.rb" "test/cases/arel/delete_manager_test.rb" "test/cases/arel/factory_methods_test.rb" "test/cases/arel/insert_manager_test.rb" "test/cases/arel/nodes/and_test.rb" "test/cases/arel/nodes/as_test.rb" "test/cases/arel/nodes/ascending_test.rb" "test/cases/arel/nodes/bin_test.rb" "test/cases/arel/nodes/binary_test.rb" "test/cases/arel/nodes/bind_param_test.rb" "test/cases/arel/nodes/case_test.rb" "test/cases/arel/nodes/casted_test.rb" "test/cases/arel/nodes/comment_test.rb" "test/cases/arel/nodes/count_test.rb" "test/cases/arel/nodes/delete_statement_test.rb" "test/cases/arel/nodes/descending_test.rb" "test/cases/arel/nodes/distinct_test.rb" "test/cases/arel/nodes/equality_test.rb" "test/cases/arel/nodes/extract_test.rb" "test/cases/arel/nodes/false_test.rb" "test/cases/arel/nodes/grouping_test.rb" "test/cases/arel/nodes/infix_operation_test.rb" "test/cases/arel/nodes/insert_statement_test.rb" "test/cases/arel/nodes/named_function_test.rb" "test/cases/arel/nodes/node_test.rb" "test/cases/arel/nodes/not_test.rb" "test/cases/arel/nodes/or_test.rb" "test/cases/arel/nodes/over_test.rb" "test/cases/arel/nodes/select_core_test.rb" "test/cases/arel/nodes/select_statement_test.rb" "test/cases/arel/nodes/sql_literal_test.rb" "test/cases/arel/nodes/sum_test.rb" "test/cases/arel/nodes/table_alias_test.rb" "test/cases/arel/nodes/true_test.rb" "test/cases/arel/nodes/unary_operation_test.rb" "test/cases/arel/nodes/update_statement_test.rb" "test/cases/arel/nodes/window_test.rb" "test/cases/arel/nodes_test.rb" "test/cases/arel/select_manager_test.rb" "test/cases/arel/table_test.rb" "test/cases/arel/update_manager_test.rb" "test/cases/arel/visitors/dispatch_contamination_test.rb" "test/cases/arel/visitors/dot_test.rb" "test/cases/arel/visitors/mysql_test.rb" "test/cases/arel/visitors/postgres_test.rb" "test/cases/arel/visitors/sqlite_test.rb" "test/cases/arel/visitors/to_sql_test.rb" "test/cases/associations/belongs_to_associations_test.rb" "test/cases/associations/bidirectional_destroy_dependencies_test.rb" "test/cases/associations/callbacks_test.rb" "test/cases/associations/cascaded_eager_loading_test.rb" "test/cases/associations/eager_load_includes_full_sti_class_test.rb" "test/cases/associations/eager_load_nested_include_test.rb" "test/cases/associations/eager_singularization_test.rb" "test/cases/associations/eager_test.rb" "test/cases/associations/extension_test.rb" "test/cases/associations/has_and_belongs_to_many_associations_test.rb" "test/cases/associations/has_many_associations_test.rb" "test/cases/associations/has_many_through_associations_test.rb" "test/cases/associations/has_one_associations_test.rb" "test/cases/associations/has_one_through_associations_test.rb" "test/cases/associations/inner_join_association_test.rb" "test/cases/associations/inverse_associations_test.rb" "test/cases/associations/join_model_test.rb" "test/cases/associations/left_outer_join_association_test.rb" "test/cases/associations/nested_through_associations_test.rb" "test/cases/associations/required_test.rb" "test/cases/associations_test.rb" "test/cases/attribute_decorators_test.rb" "test/cases/attribute_methods/read_test.rb" "test/cases/attribute_methods_test.rb" "test/cases/attributes_test.rb" "test/cases/autosave_association_test.rb" "test/cases/base_test.rb" "test/cases/batches_test.rb" "test/cases/binary_test.rb" "test/cases/bind_parameter_test.rb" "test/cases/boolean_test.rb" "test/cases/cache_key_test.rb" "test/cases/calculations_test.rb" "test/cases/callbacks_test.rb" "test/cases/clone_test.rb" "test/cases/coders/json_test.rb" "test/cases/coders/yaml_column_test.rb" "test/cases/collection_cache_key_test.rb" "test/cases/column_alias_test.rb" "test/cases/column_definition_test.rb" "test/cases/comment_test.rb" "test/cases/connection_adapters/adapter_leasing_test.rb" "test/cases/connection_adapters/connection_handler_test.rb" "test/cases/connection_adapters/connection_handlers_multi_db_test.rb" "test/cases/connection_adapters/connection_handlers_multi_pool_config_test.rb" "test/cases/connection_adapters/connection_handlers_sharding_db_test.rb" "test/cases/connection_adapters/merge_and_resolve_default_url_config_test.rb" "test/cases/connection_adapters/mysql_type_lookup_test.rb" "test/cases/connection_adapters/schema_cache_test.rb" "test/cases/connection_adapters/type_lookup_test.rb" "test/cases/connection_management_test.rb" "test/cases/connection_pool_test.rb" "test/cases/core_test.rb" "test/cases/counter_cache_test.rb" "test/cases/custom_locking_test.rb" "test/cases/database_configurations/hash_config_test.rb" "test/cases/database_configurations/resolver_test.rb" "test/cases/database_configurations_test.rb" "test/cases/database_selector_test.rb" "test/cases/database_statements_test.rb" "test/cases/date_test.rb" "test/cases/date_time_precision_test.rb" "test/cases/date_time_test.rb" "test/cases/defaults_test.rb" "test/cases/dirty_test.rb" "test/cases/disconnected_test.rb" "test/cases/dup_test.rb" "test/cases/enum_test.rb" "test/cases/errors_test.rb" "test/cases/explain_subscriber_test.rb" "test/cases/explain_test.rb" "test/cases/filter_attributes_test.rb" "test/cases/finder_respond_to_test.rb" "test/cases/finder_test.rb" "test/cases/fixture_set/file_test.rb" "test/cases/fixtures_test.rb" "test/cases/forbidden_attributes_protection_test.rb" "test/cases/habtm_destroy_order_test.rb" "test/cases/hot_compatibility_test.rb" "test/cases/i18n_test.rb" "test/cases/inheritance_test.rb" "test/cases/insert_all_test.rb" "test/cases/instrumentation_test.rb" "test/cases/integration_test.rb" "test/cases/invalid_connection_test.rb" "test/cases/invertible_migration_test.rb" "test/cases/json_attribute_test.rb" "test/cases/json_serialization_test.rb" "test/cases/locking_test.rb" "test/cases/log_subscriber_test.rb" "test/cases/migration/change_schema_test.rb" "test/cases/migration/change_table_test.rb" "test/cases/migration/column_attributes_test.rb" "test/cases/migration/column_positioning_test.rb" "test/cases/migration/columns_test.rb" "test/cases/migration/command_recorder_test.rb" "test/cases/migration/compatibility_test.rb" "test/cases/migration/create_join_table_test.rb" "test/cases/migration/foreign_key_test.rb" "test/cases/migration/index_test.rb" "test/cases/migration/logger_test.rb" "test/cases/migration/pending_migrations_test.rb" "test/cases/migration/references_foreign_key_test.rb" "test/cases/migration/references_index_test.rb" "test/cases/migration/references_statements_test.rb" "test/cases/migration/rename_table_test.rb" "test/cases/migration_test.rb" "test/cases/migrator_test.rb" "test/cases/mixin_test.rb" "test/cases/modules_test.rb" "test/cases/multi_db_migrator_test.rb" "test/cases/multiparameter_attributes_test.rb" "test/cases/multiple_db_test.rb" "test/cases/nested_attributes_test.rb" "test/cases/nested_attributes_with_callbacks_test.rb" "test/cases/null_relation_test.rb" "test/cases/numeric_data_test.rb" "test/cases/persistence_test.rb" "test/cases/pooled_connections_test.rb" "test/cases/prepared_statement_status_test.rb" "test/cases/primary_keys_test.rb" "test/cases/query_cache_test.rb" "test/cases/quoting_test.rb" "test/cases/readonly_test.rb" "test/cases/reaper_test.rb" "test/cases/reflection_test.rb" "test/cases/relation/delegation_test.rb" "test/cases/relation/delete_all_test.rb" "test/cases/relation/merging_test.rb" "test/cases/relation/mutation_test.rb" "test/cases/relation/or_test.rb" "test/cases/relation/predicate_builder_test.rb" "test/cases/relation/record_fetch_warning_test.rb" "test/cases/relation/select_test.rb" "test/cases/relation/update_all_test.rb" "test/cases/relation/where_chain_test.rb" "test/cases/relation/where_clause_test.rb" "test/cases/relation/where_test.rb" "test/cases/relation_test.rb" "test/cases/relations_test.rb" "test/cases/reload_models_test.rb" "test/cases/reserved_word_test.rb" "test/cases/result_test.rb" "test/cases/sanitize_test.rb" "test/cases/schema_dumper_test.rb" "test/cases/schema_loading_test.rb" "test/cases/scoping/default_scoping_test.rb" "test/cases/scoping/named_scoping_test.rb" "test/cases/scoping/relation_scoping_test.rb" "test/cases/secure_token_test.rb" "test/cases/serialization_test.rb" "test/cases/serialized_attribute_test.rb" "test/cases/statement_cache_test.rb" "test/cases/statement_invalid_test.rb" "test/cases/store_test.rb" "test/cases/strict_loading_test.rb" "test/cases/suppressor_test.rb" "test/cases/tasks/database_tasks_test.rb" "test/cases/tasks/mysql_rake_test.rb" "test/cases/tasks/postgresql_rake_test.rb" "test/cases/tasks/sqlite_rake_test.rb" "test/cases/test_databases_test.rb" "test/cases/test_fixtures_test.rb" "test/cases/time_precision_test.rb" "test/cases/timestamp_test.rb" "test/cases/touch_later_test.rb" "test/cases/transaction_callbacks_test.rb" "test/cases/transaction_isolation_test.rb" "test/cases/transactions_test.rb" "test/cases/type/adapter_specific_registry_test.rb" "test/cases/type/date_time_test.rb" "test/cases/type/integer_test.rb" "test/cases/type/string_test.rb" "test/cases/type/time_test.rb" "test/cases/type/type_map_test.rb" "test/cases/type/unsigned_integer_test.rb" "test/cases/type_test.rb" "test/cases/types_test.rb" "test/cases/unconnected_test.rb" "test/cases/unsafe_raw_sql_test.rb" "test/cases/validations/absence_validation_test.rb" "test/cases/validations/association_validation_test.rb" "test/cases/validations/i18n_generate_message_validation_test.rb" "test/cases/validations/i18n_validation_test.rb" "test/cases/validations/length_validation_test.rb" "test/cases/validations/numericality_validation_test.rb" "test/cases/validations/presence_validation_test.rb" "test/cases/validations/uniqueness_validation_test.rb" "test/cases/validations_test.rb" "test/cases/view_test.rb" "test/cases/yaml_serialization_test.rb" "test/cases/adapters/sqlite3/collation_test.rb" "test/cases/adapters/sqlite3/copy_table_test.rb" "test/cases/adapters/sqlite3/explain_test.rb" "test/cases/adapters/sqlite3/json_test.rb" "test/cases/adapters/sqlite3/quoting_test.rb" "test/cases/adapters/sqlite3/sqlite3_adapter_test.rb" "test/cases/adapters/sqlite3/sqlite3_create_folder_test.rb" "test/cases/adapters/sqlite3/statement_pool_test.rb" "test/cases/adapters/sqlite3/transaction_test.rb" --seed=29932]
/Users/kamipo/src/github.com/rails/rails/vendor/bundle/ruby/2.7.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/Users/kamipo/.rbenv/versions/2.7.0/bin/bundle:23:in `load'
/Users/kamipo/.rbenv/versions/2.7.0/bin/bundle:23:in `<main>'
Tasks: TOP => test_sqlite3 => test:sqlite3
(See full trace by running task with --trace)
```

eugeneius commented:
Using @kamipo's seed, I sent the logs from one of the failing tests to STDOUT to investigate:

```patch
diff --git a/activerecord/test/cases/associations/has_one_associations_test.rb b/activerecord/test/cases/associations/has_one_associations_test.rb
index f9cbfebe0e..aba145ed1d 100644
--- a/activerecord/test/cases/associations/has_one_associations_test.rb
+++ b/activerecord/test/cases/associations/has_one_associations_test.rb
@@ -810,12 +810,15 @@ class DestroyByParentAuthor < ActiveRecord::Base
   end

   test "destroyed_by_association set in child destroy callback on parent destroy" do
+    old_logger, ActiveRecord::Base.logger = ActiveRecord::Base.logger, Logger.new(STDOUT)
     author = DestroyByParentAuthor.create!(name: "Test")
     book = DestroyByParentBook.create!(author: author)

     author.destroy

     assert_not DestroyByParentBook.exists?(book.id)
+  ensure
+    ActiveRecord::Base.logger = old_logger
   end

   test "destroyed_by_association set in child destroy callback on replace" do
```

```
D, [2020-05-04T20:34:07.808055 #70833] DEBUG -- :   TRANSACTION (0.1ms)  SAVEPOINT active_record_1
D, [2020-05-04T20:34:07.808605 #70833] DEBUG -- :   HasOneAssociationsTest::DestroyByParentAuthor Create (0.4ms)  INSERT INTO "authors" ("name") VALUES (?)  [["name", "Test"]]
D, [2020-05-04T20:34:07.808834 #70833] DEBUG -- :   TRANSACTION (0.1ms)  RELEASE SAVEPOINT active_record_1
D, [2020-05-04T20:34:07.825359 #70833] DEBUG -- :   TRANSACTION (0.1ms)  SAVEPOINT active_record_1
D, [2020-05-04T20:34:07.825886 #70833] DEBUG -- :   HasOneAssociationsTest::DestroyByParentBook Create (0.3ms)  INSERT INTO "books" ("author_id", "created_at", "updated_at", "updated_on") VALUES (?, ?, ?, ?)  [["author_id", 4], ["created_at", "2020-05-04 19:34:07.824921"], ["updated_at", "2020-05-04 19:34:07.824921"], ["updated_on", "2020-05-04"]]
D, [2020-05-04T20:34:07.826119 #70833] DEBUG -- :   TRANSACTION (0.0ms)  RELEASE SAVEPOINT active_record_1
D, [2020-05-04T20:34:07.827155 #70833] DEBUG -- :   TRANSACTION (0.0ms)  SAVEPOINT active_record_1
D, [2020-05-04T20:34:07.827411 #70833] DEBUG -- :   HasOneAssociationsTest::DestroyByParentBook Load (0.1ms)  SELECT "books".* FROM "books" WHERE "books"."author_id" = ? LIMIT ?  [["author_id", 4], ["LIMIT", 1]]
D, [2020-05-04T20:34:07.828146 #70833] DEBUG -- :   HasOneAssociationsTest::DestroyByParentBook Destroy (0.1ms)  DELETE FROM "books" WHERE "books"."id" = ?  [["id", 5]]
D, [2020-05-04T20:34:07.828911 #70833] DEBUG -- :   HasOneAssociationsTest::DestroyByParentAuthor Destroy (0.2ms)  DELETE FROM "authors" WHERE "authors"."id" = ?  [["id", 4]]
D, [2020-05-04T20:34:07.829205 #70833] DEBUG -- :   TRANSACTION (0.1ms)  RELEASE SAVEPOINT active_record_1
D, [2020-05-04T20:34:07.830778 #70833] DEBUG -- :   HasOneAssociationsTest::DestroyByParentBook Exists? (0.3ms)  SELECT 1 AS one FROM "books" WHERE "books"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
```

Where did 6 come from? 🤔 More logging:

```patch
diff --git a/activerecord/test/cases/associations/has_one_associations_test.rb b/activerecord/test/cases/associations/has_one_associations_test.rb
index aba145ed1d..7dcc9a9fbd 100644
--- a/activerecord/test/cases/associations/has_one_associations_test.rb
+++ b/activerecord/test/cases/associations/has_one_associations_test.rb
@@ -814,6 +814,8 @@ class DestroyByParentAuthor < ActiveRecord::Base
     author = DestroyByParentAuthor.create!(name: "Test")
     book = DestroyByParentBook.create!(author: author)

+    puts "number of books: #{DestroyByParentBook.where(author: author).count}"
+
     author.destroy

     assert_not DestroyByParentBook.exists?(book.id)
```

```
number of books: 2
```

Hmm...

```patch
diff --git a/activerecord/test/cases/associations/has_one_associations_test.rb b/activerecord/test/cases/associations/has_one_associations_test.rb
index 40bef912e7..8137440d43 100644
--- a/activerecord/test/cases/associations/has_one_associations_test.rb
+++ b/activerecord/test/cases/associations/has_one_associations_test.rb
@@ -810,12 +810,12 @@ class DestroyByParentAuthor < ActiveRecord::Base
   end

   test "destroyed_by_association set in child destroy callback on parent destroy" do
+    puts "number of books: #{DestroyByParentBook.where(author_id: 4).count}"
+
     old_logger, ActiveRecord::Base.logger = ActiveRecord::Base.logger, Logger.new(STDOUT)
     author = DestroyByParentAuthor.create!(name: "Test")
     book = DestroyByParentBook.create!(author: author)

-    puts "number of books: #{DestroyByParentBook.where(author: author).count}"
-
     author.destroy

     assert_not DestroyByParentBook.exists?(book.id)
```

```
number of books: 1
```

It looks like the database isn't being cleaned up properly after a previous test.

I'll continue to poke at this, but wanted to checkpoint what I've discovered so far.

jonathanhefner commented:
@eugeneius Good catch!  I was just discovering this too!  I added `DestroyByParentBook.delete_all` to the beginning of the test case and it "fixed" the failure.  Obviously that's not a proper fix though.

eugeneius commented:
Ah! https://github.com/rails/rails/pull/39124 added a test to transactions_test.rb, which [disables `use_transactional_tests`](https://github.com/rails/rails/blob/8d02c0d300c12cbb2d6b2060ed5e59a466561f9f/activerecord/test/cases/transactions_test.rb#L14), so the database isn't cleaned up afterwards. I still can't tell how author_id=4 shows up again though: is another test resetting the auto increment counter?

Cleaning up in a `teardown` hook, as we do [elsewhere](https://github.com/rails/rails/blob/8d02c0d300c12cbb2d6b2060ed5e59a466561f9f/activerecord/test/cases/autosave_association_test.rb#L835-L836), might be a reasonable solution here.

eugeneius commented:
author_id=4 shows up again because there are [three author fixtures](https://github.com/rails/rails/blob/8d02c0d300c12cbb2d6b2060ed5e59a466561f9f/activerecord/test/fixtures/authors.yml), so tests that use those fixtures will always get id=4 for the first author that they create.

This test file doesn't actually depend on those fixtures directly, and only started to load them in https://github.com/rails/rails/pull/30103 to address flakiness in these very same tests!

The simplest solution seems to be to load the books fixtures in this test file, so that the leftover state in that table is also cleaned up. Ideally we would always clean up manually after disabling transactional tests, but that's hard to guarantee, as I demonstrated in https://github.com/rails/rails/pull/39124 by failing to do it. 😄

jonathanhefner commented:
> The simplest solution seems to be to load the books fixtures in this test file, so that the leftover state in that table is also cleaned up.

Agreed!  I'll modify this PR and you and Ryuta as co-authors.  

> Ideally we would always clean up manually after disabling transactional tests, but that's hard to guarantee, as I demonstrated in #39124 by failing to do it.

There does seem to be some insidious behavior here that would be nice to fix.  When a test class turns off `use_transactional_tests`, it's expected that its own methods can trip each other up.  But, I think it's a bit of a foot-gun to allow it to trip up other test classes.  I wonder if it would be feasible to wrap all tests (not each test) of a test class in a transaction when `use_transactional_tests` is disabled?

eugeneius commented:
Test cases typically disable transactional tests when having an open transaction would invalidate the test scenario; transactions_test.rb is a good example, since it needs to test Rails' handling of top-level transactions, not just savepoints. I agree that ideally this would "just work", though.

We had this same problem in our application, and solved it by invoking the database_cleaner gem after every test that runs with transactional tests disabled. That exact approach won't fly here, but perhaps [truncating all the things](https://github.com/rails/rails/blob/8d02c0d300c12cbb2d6b2060ed5e59a466561f9f/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L187) could be a reasonable substitute? Worth exploring as a followup.

jonathanhefner committed:
Fix flakey destroyed_by_association tests

Example failure: https://buildkite.com/rails/rails/builds/68661#84f8790a-fc9e-42ef-a7fb-5bd15a489de8/1002-1012

The failing `destroyed_by_association` tests create an author (a
DestroyByParentAuthor) and a book (a DestroyByParentBook) that belongs
to that author.  If the database already contains books that refer to
that author's ID from previous tests (i.e. tests that disabled
`use_transactional_tests`), then one of those books will be loaded and
destroyed instead of the intended DestroyByParentBook book.

By loading the `:books` fixtures, we ensure the database does not
contain such unexpected books.

Co-authored-by: Eugene Kenny <elkenny@gmail.com>
Co-authored-by: Ryuta Kamizono <kamipo@gmail.com>

eugeneius commented:
Thanks @jonathanhefner! 🙌

jonathanhefner commented:
Thank *you* @eugeneius and @kamipo! :heart:
